spec.bs

```diff
@@ -5863,6 +5832,204 @@ an [=origin=] |topLevelOrigin|, and a [=string=] |slotSizeQueryParam|:
 
 </div>
 
+<h3 id=trusted-seller-signals-batcher>Trusted scoring signals batcher</h3>
+
+A <dfn>trusted scoring signals batcher</dfn> helps manage merging multiple trusted scoring signals
+requests into smaller number of fetches. It's a [=struct=] with the following [=struct/items=]:
+
+<dl dfn-for="trusted scoring signals batcher">
+  : <dfn>request queue</dfn>
+  :: A [=list=] of [=trusted scoring signals requests=] that hasn't yet been organized to aid
+    batching.
+  : <dfn>request map</dfn>
+  :: A [=map=] from a tuple of [=script fetcher=], [=URL=], {{unsigned short}} or null, and

```

Comment:
We create a new batcher for each fetcher, though, so the fetcher is always the same, and could be a top-level pointer.

Reply:
Yeah, but this removes the decision from here?

Reply:
I mean, it already is there, so I'm wondering why it's here as well?  Do you plan to change things to use a global batcher?  If we're partitioning by script fetcher anyways, I'm not sure what that would get us?

Reply:
To make sure we're on the same page:  We create a scoring signals fetcher, then we create a batcher on the next line.  We only ever use that batcher with that fetcher.  So we could just pass the fetch to the batcher on batcher construction, and not key each entry in bather's map on the fetcher.  It's not clear to me why we aren't doing that.

Reply:
I want to retain the ability to change this relation.

Reply:
If we don't share fetches when the script fetcher is different, I'm not sure that gets us anything, though?

Reply:
We could make the batcher global (doesn't V2 require that for the cache?) and then maybe have the fetcher per auction, or have some optional way for fetchers to be shared (which is what impl does if there are concurrent auctions)?

Reply:
V2 does indeed use a global batcher.  In V2, I also have no plans to partition anything based on the fetcher, just on the seller origin (and main frame origin, and URL, and doubtless a couple other things).  Guess perhaps I should have led with that.

Reply:
impl-wise, the partitioning by script fetcher roughly corresponds to different script URLs getting different worklet processes, and in cross-origin case those waiting for script headers before asking trusted signals for stuff. I guess in spec those two would eventually need to be separate stages of the batching process, with cross-origin things first waiting for the headers before being considered for merging, but that doesn't seem to be necessary yet.

The spec just does a bad job of representing this sort of thing at all; the previous CL you reviewed changed "we fetch the script every single time we call scoreAd" to "we fetch it once per component auction", which is closer but not 100%... and re-arranging lifetime of ScriptFetchers is basically about that.

(One of my TODO entries is doing that for bid scripts, since those still fetch for every generateBid()).

Reply:
Different script URLs share worklet processes, just not objects within the process.

Anyhow, I'm OK with this as-is, just wanted to dive into it a bit.  Should finish reviewing this by EOD.